feat(widgets): implement Widget for Widget refs

joshka opened the pull request:
Many widgets can be rendered without changing their state.

This commit implements The `Widget` trait for references to
widgets and changes their implementations to be immutable.

This allows us to render widgets without consuming them by passing a ref
to the widget when calling `Frame::render_widget()`.

```rust
// this might be stored in a struct
let paragraph = Paragraph::new("Hello world!");

let [left, right] = area.split(&Layout::horizontal([20, 20]));
frame.render_widget(&paragraph, left);
frame.render_widget(&paragraph, right); // we can reuse the widget
```

Implemented for all widgets except BarChart (which has an implementation that
modifies the internal state and requires a rewrite to fix.

Other widgets will be implemented in follow up commits.

Fixes: https://github.com/ratatui-org/ratatui/discussions/164
Replaces PRs: https://github.com/ratatui-org/ratatui/pull/122 and https://github.com/ratatui-org/ratatui/pull/16
Enables: https://github.com/ratatui-org/ratatui/issues/132
Validated as a viable working solution by: https://github.com/ratatui-org/ratatui/pull/836

joshka commented:
Not all the examples have been updated, but this is about a good enough size to review as it stands.

Valentin271 reviewed (approved):
As stated on discord I like this way of introducing render by reference. Approved!

joshka commented:
Fun fact that I discovered earlier tonight.

`RefWidget` and `MutWidget` aren't actually necessary as the same effect can be had by implementing `Widget` on `&T`.

```rust
impl Widget for Block<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        (&self).render(area, buf);
    }
}

impl Widget for &Block<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        if area.is_empty() {
            return;
        }
        self.render_borders(area, buf);
        self.render_titles(area, buf);
    }
}
```
This allows you to do all the things you'd need to against a reference to the widget - you can store it, render a reference to it, etc.
```rust
let block = Block::default().title("Block").borders(Borders::ALL);

// does not compile
terminal.draw(|frame| {
    frame.render_widget(block, frame.size());
    let _area = block.inner(Rect::default());
})?;

// compiles
terminal.draw(|frame| {
    frame.render_widget(&block, frame.size());
    let _area = block.inner(Rect::default());
})?;
```

The upshot of this is that instead of adding two traits, we should instead implement Widget on &T more often.
The latest commit makes this change and removes the RefWidget and MutWidget traits.

kdheepak commented:
And instead of MutWidget we continue using the stateful widget strategy? I actually like that a lot.

orhun commented:
Seconding kd here, I think having `impl &T` is much cleaner solution. And yeah, nice find!

joshka commented:
>And instead of MutWidget we continue using the stateful widget strategy? I actually like that a lot.

`impl Widget on &mut SomeWidget` also works for situations where it makes sense to just consider the widget as the state instead of having external state.

What this does suggest is that we could do with some really clear docs on these scenarios.

Valentin271 commented:
+1 for implementing on `&T` and `&mut T`.
Also +1 for keeping the `StatefulWidget` pattern.

> What this does suggest is that we could do with some really clear docs on these scenarios.

Yep if we keep `StatefuleWidget` and `&mut T` this is essential.

joshka commented:
`StatefulWdget` shouldn't go anywere - it's useful to have the state and widget separate - mainly it simplifies the application state. Often state values won't have any lifetimes, as they're just storing indexes, etc. This simplification is helpful to newer users (as it avoids the pain of adding lifetimes everywhere).

joshka commented on tests/widgets_calendar.rs:
```diff
@@ -63,7 +64,7 @@ fn show_month_header() {
     )
     .show_month_header(Style::default());
     let expected = Buffer::with_lines(vec![
-        "     January 2023    ",
+        "    January 2023     ",
```
This change is due to changing the implementation to use `Line::render()` instead of manually calculating the position.

When centering text, Ratatui will generally center 1 space to the left if the centering isn't exact instead of 1 space to the right.

joshka commented:
I rebased this on main, squashed to remove the RefWidget implmentation, and updated the commit message and PR comment to fit as a good changelog message.

Valentin271 commented on src/widgets/block.rs:
```diff
@@ -703,13 +721,24 @@ impl<'a> Block<'a> {
-            return;
+/// An extension trait for [`Block`] that provides some convenience methods.
+///
+/// This is pub(crate) for now because it is not clear if this is the best way to do this.
+pub(crate) trait BlockExt {
```
This seems like a good idea for now at least!

joshka replied:
I'm not 100% sure I like the render and modify Rect being tied always. that could be annoying for a caller that wants to do something with the original area. (Example: put a scrollbar on the block border instead of within the border).

Making this pub(crate) makes it easy to change this without breaking user code.

Valentin271 replied:
Maybe the name `render` should just be changed to `render_option` or similar. To avoid confusion and be able to switch render easily.

joshka replied:
That could work. Maybe we could also add the following usefully:

```rust
fn inner() -> Rect { ... }
```

and split the render call into two (this kinda looks a little oddish):
```rust
self.block.render_optional(area, buf);
let area = self.block.inner_optional();
```

Valentin271 replied:
> ```rust
> self.block.render_optional(area, buf);
> let area = self.block.inner_optional();
> ```

I didn't like this approach at first but changing `_optional` to`_if_exists` feels more descriptive to me.

```rust
self.block.render_if_exists(area, buf);
let area = self.block.inner_if_exists();
```

joshka replied:
I decided to split this and leave it as render(), it seems to look better with two calls (render, inner) rather than one that mutates the area.

Valentin271 commented on tests/widgets_calendar.rs:
```diff
@@ -11,6 +11,7 @@ use ratatui::{
 };
 use time::{Date, Month};
 
+#[track_caller]
```
I just discovered this, why is it needed here?

joshka replied:
When the test fails, it should report the line of the failing test, and not the line in this method.
This makes things easier to fix when failures occur in this method.

Valentin271 replied:
Alright interesting, thanks!

Valentin271 reviewed (commented):
Looks good, only the doc is a mandatory change for me.

joshka commented on src/widgets.rs:
```diff
@@ -53,7 +53,24 @@ pub use self::{
-/// Base requirements for a Widget
+/// A `Widget` is a type that can be drawn on a [`Buffer`] in a given [`Rect`].
+///
+/// Widgets are created for each frame as they are consumed after rendered. They are not meant to be
+/// stored but used as *commands* to draw common figures in the UI.
```
Ha yup - needs a reword. I think this was GPT-ibberish.

joshka replied:
Updated in current

joshka commented:
I've pulled in the changes from main and finished off all the widgets except BarChart.

joshka commented on src/prelude.rs:
```diff
@@ -31,4 +31,5 @@ pub use crate::{
     symbols::{self, Marker},
     terminal::{CompletedFrame, Frame, Terminal, TerminalOptions, Viewport},
     text::{self, Line, Masked, Span, Text},
+    widgets::{block::BlockExt, StatefulWidget, Widget},
```
I think it makes sense to add StatefulWidget and Widget to the prelude. Wvery app that uses Ratatui calls render on one or more widgets, and these are not likely to cause problems with existing code.

joshka commented on tests/widgets_chart.rs:
```diff
@@ -30,38 +31,36 @@ where
+#[case(1, 0)]
+#[case(1, 1)]
+#[case(2, 2)]
+fn widgets_chart_can_render_on_small_areas(#[case] width: u16, #[case] height: u16) {
```
the reason for this change was that I broke this test when changing the chart, and it was annoying to work out which test case broke. I fixed the break, but left the improved test.

joshka commented on src/widgets/barchart.rs:
```diff
@@ -36,6 +34,9 @@ use super::{Block, Widget};
 /// The chart can have a [`Direction`] (by default the bars are [`Vertical`](Direction::Vertical)).
 /// This is set using [`BarChart::direction`].
 ///
+/// Note: this is the only widget that doesn't implement `Widget` for `&T` because the current
+/// implementation modifies the internal state of self. This will be fixed in the future.
+///
```
The barchart is updated when it renders by removing bars that are not visible. If we implemented Widget on @mut Barchart, then the state is changed forever. If the application shrinks the visible area and then grows it (due to screen resize or application layout changes, the removed bars and groups are lost forever.

This can be fixed, but it's not as straightforward as all the other changes which were mostly just repeating the same pattern. The fix is to make the BarChart immutable. An observation on this is that Widgets shouldn't generally mutate parts of their state in destructive ways.

See comments on the implementation of this starting at https://github.com/ratatui-org/ratatui/pull/288#issuecomment-1632466975 for some more context, and also on a followup PR https://github.com/ratatui-org/ratatui/pull/532#pullrequestreview-1643340812 particularly note this comment:

> The remove invisible groups and bars approach (which mutates the state) works because the widget is consumed on render. If we change the widget to be borrowed in the future, then the entire logic for the widget needs to rewritten (this is something that we're considering).

Valentin271 replied:
Ok makes sense, I should have checked beforehand. :+1: Lets open an issue for that.

Valentin271 replied:
Issue at https://github.com/ratatui-org/ratatui/issues/870.

joshka commented on src/widgets/block.rs:
```diff
@@ -703,13 +717,31 @@ impl<'a> Block<'a> {
+}
+
+impl BlockExt for Option<Block<'_>> {
+    fn render(&self, area: Rect, buf: &mut Buffer) {
```
Makes sense - I think what I'd tried to do was implement the render with the update area function on Option<Block>, which can't work, but now that these are split, I can do this fine. In fact we can go one step further than this and implement `Widget` for all `Option<W: Widget>`

joshka replied:
I'm wrong - I can't get this to work with Option<W>

joshka replied:
I made this work

joshka commented on src/widgets/block.rs:
```diff
@@ -703,13 +717,31 @@ impl<'a> Block<'a> {
-        self.render_titles(area, buf);
+    }
+
+    fn inner(&self, area: Rect) -> Rect {
```
Can you point to an example or two in std for this?

Valentin271 replied:
One example would be https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_and. But that's actually the only relevant one I can find.

I was thinking of methods like `map_err` and similar but this is actually on `Result` only. I though `Option` had similar names like `map_some`.

joshka replied:
Went with inner_if_some.

joshka commented on src/widgets.rs:
````diff
@@ -53,7 +53,29 @@ pub use self::{
+///     // A widget can be rendered by simply calling its `render` method.
+///     frame.render_widget(Clear, frame.size());
+/// });
+/// ```
 pub trait Widget {
````
added

Valentin271 reviewed (approved):
All good to me. We'll finally have render by ref! I still can't believe it was this simple.
